File: Difficulty/B. Medium/3254-find-the-power-of-k-size-subarrays-i/3254-find-the-power-of-k-size-subarrays-i.c
#include <stdlib.h>

/**
 * @brief Computes the power of all k-size subarrays of nums efficiently.
 *
 * @param nums (int*): Array of integers.
 * @param numsSize (int): Size of the nums array.
 * @param k (int): Size of subarrays to process.
 * @param returnSize (int*): Pointer to store the size of the returned array.
 *
 * @return (int*): Dynamically allocated array containing the results.
 *                 Caller must free the returned array.
 */
int* resultsArray(int* nums, int numsSize, int k, int* returnSize) {
    // Handle the edge case when k == 1
    if (k == 1) {
        *returnSize = numsSize;  // Each element is its own subarray
        int* results = (int*)malloc(numsSize * sizeof(int));
        if (!results) return NULL;  // Return NULL if memory allocation fails

        // Copy all elements from nums to results
        for (int i = 0; i < numsSize; i++) {
            results[i] = nums[i];
        }
        return results;
    }

    *returnSize = numsSize - k + 1;  // Size of the result array
    int* results = (int*)malloc((*returnSize) * sizeof(int));  // Allocate memory for the result array
    if (!results) {
        return NULL;  // Return NULL if memory allocation fails
    }

    int left = 0, right = 1;  // Sliding window pointers
    int idx = 0;  // Index for the results array

    while (right < numsSize) {
        if (nums[right] - nums[right - 1] != 1) {
            // If the current pair is not consecutive
            while (left < right && left + k - 1 < numsSize) {
                results[idx++] = -1;  // Fill invalid windows with -1
                left++;
            }
            left = right;  // Reset left pointer to the current right position
        } else if (right - left == k - 1) {
            // If the window is valid and its size is exactly k
            results[idx++] = nums[right];  // Append the last element of the valid window
            left++;  // Move the left pointer forward
        }
        right++;  // Increment the right pointer
    }

    return results;  // Return the dynamically allocated results array
}

File: Difficulty/B. Medium/3254-find-the-power-of-k-size-subarrays-i/3254-find-the-power-of-k-size-subarrays-i.py
from typing import List

class Solution:
    def resultsArray(self, nums: List[int], k: int) -> List[int]:
        """
        Computes the power of all k-size subarrays of nums efficiently.

        Parameters:
        nums (List[int]): Array of integers.
        k (int): Size of subarrays to process.

        Returns:
        List[int]: List of powers for each subarray of size k.
        """
        if k == 1: 
            return nums  # If k == 1, each element is its own subarray and the result is nums

        results = []  # List to store the results
        left, right = 0, 1  # Pointers for the sliding window
        n = len(nums)

        # Iterate through the array using the right pointer
        while right < n:
            # If the current pair is not consecutive
            if nums[right] - nums[right - 1] != 1:
                # Fill -1 for all invalid windows starting from left
                while left < right and left + k - 1 < n:
                    results.append(-1)
                    left += 1
                # Reset left pointer to the current right position
                left = right
            # If the window is valid and its size is exactly k
            elif right - left == k - 1:
                results.append(nums[right])  # Append the last element of the valid window
                left += 1  # Move the left pointer forward

            # Increment the right pointer
            right += 1

        return results
